File: python/src/tdcv2/prng/rand.py
"""Drawing from a generator: an integer, an element, a shuffle.

Small on purpose, and shared rather than reimplemented per generator: the ORDER in which draws
are taken is part of the output. A second copy of "pick an element" that consumed a different
number of draws would silently shift every value after it.
"""

from __future__ import annotations

import math
from typing import TypeVar

from .prng import Sfc32

T = TypeVar("T")


def next_int(prng: Sfc32, minimum: int, maximum: int) -> int:
    """An integer in ``[minimum, maximum)``. The caller guarantees the range is non-empty."""
    return math.floor(prng.next() * (maximum - minimum) + minimum)


def pick(prng: Sfc32, items: list[T]) -> T:
    """A uniformly chosen element of a non-empty list."""
    return items[math.floor(prng.next() * len(items))]


def shuffle(prng: Sfc32, items: list[T]) -> list[T]:
    """Fisher-Yates, returning a new list and leaving the input untouched."""
    out = list(items)
    for i in range(len(out) - 1, 0, -1):
        j = math.floor(prng.next() * (i + 1))
        out[i], out[j] = out[j], out[i]
    return out

File: java/src/main/java/io/github/nickliapin/tdc/prng/Rand.java
package io.github.nickliapin.tdc.prng;

import java.util.ArrayList;
import java.util.List;

/**
 * Helpers layered on a {@link Prng.Sfc32}. They hold no state of their own; every draw comes
 * from the generator passed in, so the number and order of draws is what makes a run
 * reproducible.
 *
 * <p>The shuffle in particular has to match the reference exactly: it walks from the end and
 * swaps with a random earlier-or-equal index. Walking the other direction, or drawing before
 * the swap instead of after, produces a different permutation from the same seed.
 */
public final class Rand {

  private Rand() {}

  /** An integer in [min, max). The caller guarantees {@code min < max}. */
  public static int nextInt(Prng.Sfc32 prng, int min, int max) {
    return (int) Math.floor(prng.next() * (max - min) + min);
  }

  /** A uniformly chosen element of a non-empty list. */
  public static <T> T pick(Prng.Sfc32 prng, List<T> items) {
    return items.get((int) Math.floor(prng.next() * items.size()));
  }

  /** Fisher-Yates, returning a new list and leaving the input untouched. */
  public static <T> List<T> shuffle(Prng.Sfc32 prng, List<T> items) {
    List<T> out = new ArrayList<>(items);
    for (int i = out.size() - 1; i > 0; i--) {
      int j = (int) Math.floor(prng.next() * (i + 1));
      T tmp = out.get(i);
      out.set(i, out.get(j));
      out.set(j, tmp);
    }
    return out;
  }
}
